Usage Page and Usage on Linux with hidraw and whitespace cleanup

haata opened the pull request:
Adapted from ApeironTuska's PR to PastaJ36/hidapi (https://github.com/PastaJ36/hidapi/pull/1)
Which was adapted from djpnewton's PR to signal11/hidapi (https://github.com/signal11/hidapi/pull/6)

Also addresses some of the issues mentioned in (https://github.com/signal11/hidapi/pull/6)
* hid_open_path and hid_enumerate both need to retrieve the usage page
as the user may call hid_open_path directly without using hid_enumerate
* Added get_hid_item_size() for hid parsing, used in both uses_numbered_reports() and get_usage()

NOTE: This commit does not handle composite HID descriptors
I am interested in adding support for composite descriptors though I still need to find a
device with a composite descriptor to test it correctly

- hidtest-hidraw test -

Device Found
  type: 308f 0015
  path: /dev/hidraw10
  serial_number: 53373100323943353230353139363032 - sam4s2
  Manufacturer: Kiibohd
  Product:      Keyboard - None PartialMap USBxUART
  Release:      4ac
  Interface:    0
  Usage (page): 0x6 (0x1)

Device Found
  type: 308f 0015
  path: /dev/hidraw11
  serial_number: 53373100323943353230353139363032 - sam4s2
  Manufacturer: Kiibohd
  Product:      Keyboard - None PartialMap USBxUART
  Release:      4ac
  Interface:    1
  Usage (page): 0x6 (0x1)

Device Found
  type: 308f 0015
  path: /dev/hidraw12
  serial_number: 53373100323943353230353139363032 - sam4s2
  Manufacturer: Kiibohd
  Product:      Keyboard - None PartialMap USBxUART
  Release:      4ac
  Interface:    2
  Usage (page): 0x1 (0xc)

Device Found
  type: 308f 0015
  path: /dev/hidraw13
  serial_number: 53373100323943353230353139363032 - sam4s2
  Manufacturer: Kiibohd
  Product:      Keyboard - None PartialMap USBxUART
  Release:      4ac
  Interface:    3
  Usage (page): 0x2 (0x1)

Device Found
  type: 308f 0015
  path: /dev/hidraw14
  serial_number: 53373100323943353230353139363032 - sam4s2
  Manufacturer: Kiibohd
  Product:      Keyboard - None PartialMap USBxUART
  Release:      4ac
  Interface:    4
  Usage (page): 0x1100 (0xff1c)

Youw commented:
Thanks you very much for doing this PR.
We're looking forward to accept the changes.

Before I go deep into the changes, a few general comments:

1) Please remove all white-space changes: they are irrelevant to the change itself and makes harder to look thru history of changes, e.g. it took me some time to figure that only 2 files had been changed, not 4;

2) `get_bytes` is extremely unfriendly name for a function, that actually interpret array of bytes as big-endian number;

haata commented:
Sounds good, I'll update the commit shortly.

haata commented:
Updated the commit.

- Removed the whitespace changes
- Changed `get_bytes` to `get_hid_report_bytes` and added `_hid` to get `get_hid_item_size`, `get_hid_usage` and `get_hid_report_descriptor` to make it more consistent. I've left `uses_numbered_reports` as it's a very hid specific thing (and would require a bunch more renaming to make consistent).

haata commented:
Also, for reference it seems that in order for the Linux kernel to separate composite descriptors into multiple devices as multiple udev nodes a kernel patch (per device) is required to add `HID_QUIRK_MULTI_INPUT` to https://github.com/torvalds/linux/blob/master/drivers/hid/hid-quirks.c

https://electronics.stackexchange.com/a/400268

Given that there are only a handful of devices that support this (and there doesn't seem to be any additional udev flags to help identify the order in the descriptor), I don't think any additional code for Composite HID descriptors will be useful in Linux without a bunch of kernel work (and any hidraw code will have to know how to deal with older and newer kernels).

Possibly one change that could help would be retrieving a list of all of the usages and usage pages from the descriptor and providing a list in Linux for situations where matching is more important.

haata commented:
I was able to add support for Bluetooth Usage Page and Usage.

```
Device Found
  type: 093a 2801
  path: /dev/hidraw12
  serial_number: DA:33:7F:D0:17:40
  Manufacturer:
  Product:      Hexgears-BK2
  Release:      0
  Interface:    -1
  Usage (page): 0x6 (0x1)
```

haata commented:
I believe I have per-usage page/usage pair working with hidraw. The implementation is a bit cleaner.

The device listing should match to what macOS currently does with hidapi.

z3ntu commented:
I'm getting a different behavior as root and as non-root:

normal user:
```
Device Found
  type: 1532 006c
  path: /dev/hidraw11
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    0
  Usage (page): 0x0 (0x0)

Device Found
  type: 1532 006c
  path: /dev/hidraw12
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    1
  Usage (page): 0x0 (0x0)

Device Found
  type: 1532 006c
  path: /dev/hidraw13
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    2
  Usage (page): 0x0 (0x0)
```

With sudo:
```
Device Found
  type: 1532 006c
  path: /dev/hidraw11
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)

Device Found
  type: 1532 006c
  path: /dev/hidraw11
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0x1)

Device Found
  type: 1532 006c
  path: /dev/hidraw12
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    1
  Usage (page): 0x6 (0x1)

Device Found
  type: 1532 006c
  path: /dev/hidraw12
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xc)

Device Found
  type: 1532 006c
  path: /dev/hidraw12
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    1
  Usage (page): 0x80 (0x1)

Device Found
  type: 1532 006c
  path: /dev/hidraw12
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    1
  Usage (page): 0x0 (0x1)

Device Found
  type: 1532 006c
  path: /dev/hidraw12
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    1
  Usage (page): 0x0 (0x1)

Device Found
  type: 1532 006c
  path: /dev/hidraw13
  serial_number: 
  Manufacturer: Razer
  Product:      Razer Mamba Elite
  Release:      200
  Interface:    2
  Usage (page): 0x6 (0x1)
```

The second to last and third to last entries seem to be identical though - at least from the information printed by hidtest. With one of those removed the output seems to match the macOS version.

haata commented:
You must have device permissions in order to read from hidraw. Either by setting up udev or using root.

haata commented:
@z3ntu Can you include the full hid descriptor here? I can try to debug the issue.

z3ntu commented:
@haata Can you give me a command to get this descriptor please? :)

haata commented:
I've used this utility in the past, not sure if it's available on your distro.
https://github.com/DIGImend/usbhid-dump

The last example will convert the hid descriptor to human readable format.

z3ntu commented:
Is this good?
```
$ sudo ./src/usbhid-dump --model=1532:006C
001:028:002:DESCRIPTOR         1586977867.981403
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

001:028:001:DESCRIPTOR         1586977867.982019
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 0C 09 01 A1 01 85
 02 19 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81
 00 75 08 95 0D 81 01 C0 05 01 09 80 A1 01 85 03
 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95 05
 81 01 75 08 95 0E 81 01 C0 05 01 09 00 A1 01 85
 04 09 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95
 0F 81 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00
 26 FF 00 35 00 46 FF 00 75 08 95 0F 81 00 C0

001:028:000:DESCRIPTOR         1586977867.982856
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 06 00
 FF 09 02 15 00 25 01 75 08 95 5A B1 01 C0

```

Youw commented:
> Also, is creating a new device per usage part of the API?

This is the default and only behavior on Windows. Best approach to make the API consistent - implement the wider behavior on all platforms

> Why can't we just provide an array with the usages?

1) This is a big API change. HIDAPI has a very stable API and many applications that rely on it. Developer can't just tell to all library users, that they need to change their implementation

2) On Window each usage is reported as a separate device

FFY00 commented:
> > Also, is creating a new device per usage part of the API?
> 
> This is the default and only behavior on Windows. Best approach to make the API consistent - implement the wider behavior on all platforms
> 
> > Why can't we just provide an array with the usages?
> 
>     1. This is a big API change. HIDAPI has a very stable API and many applications that rely on it. Developer can't just tell to all library users, that they need to change their implementation
> 
>     2. On Window each usage is reported as a separate device

Well, that's unfortunate. If you can tie the those HID devices in the Windows API down to a physical device then you can wrap around the behavior. This could be introduced in our API without breaking changes, just ABI changes.

FFY00 commented:
Looks like it can be done with `GetPhysicalDescriptor`, so I don't see any blocker. Let me know if I am wrong.

haata commented:
I'm not sure there's a way to get the full HID descriptor on macOS (though I would be happy to be proven wrong).

Youw commented:
> Looks like it can be done with GetPhysicalDescriptor

Have you tried it? Does it work in all cases for all possible users in this library in production?

> could be introduced in our API without breaking changes

Replacing a property field with an array of values - is a very definition of breaking change

> just ABI changes

That's a breaking change too

Youw commented:
> This seems a bit silly. Why are we duplicating the code to parse the report descriptor in every backend?

When you're the only developer and the user of a project, it may sound like something that can be refactored, improved, to make your self feel better.

When hundreds of users rely on the behavior they got used to in years - introducing a new source file with a some common implementation is a big deal. In ideal world agree that t would be much better that way. In practice - I personally know a few users, who rely of the fact, that this library consist of a single `hid.c` source file.

Secondly, this project is maintained by an open-source community purely on their enthusiasm.
If you wan't to help - PRs are welcome (no breaking changes!).

> let's say I want to read the raw HID events from mice

Good luck with that. macOS and Windows already prohibit that as a security concern (because you shouldn't be allowed to sniff some passwords, user may want to type in). My opinion - it is a question of "when" not "if" Linux would prohibit it too.

haata commented:
Is there anything pending for this PR that I can help with?

Youw commented on linux/hid.c:
```diff
@@ -566,6 +734,52 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
+			/* Usage Page and Usage */
+			int res;
+			struct hidraw_report_descriptor rpt_desc;
+			int device_handle = open(dev_path, O_RDWR);
```
can you check is `O_RDONLY` enough for this?

haata replied:
O_RDONLY seems to be enough from my testing.

Youw commented on linux/hid.c:
```diff
@@ -566,6 +734,52 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
+			int res;
+			struct hidraw_report_descriptor rpt_desc;
+			int device_handle = open(dev_path, O_RDWR);
+			if (device_handle > 0) {
```
```suggestion
			if (device_handle >= 0) {
```
`0` - is a valid file descriptor and may be returned by `open`

haata replied:
Pushed the change (and rebased on top of master)

Youw reviewed (commented):
I think this is active pending actions

Youw commented:
> I'm getting a different behavior as root and as non-root:

@z3ntu 
what about after https://github.com/libusb/hidapi/pull/139/commits/ab2516fcbae280c42911f6704537ff2fe4afd409 ?

haata commented:
> > I'm getting a different behavior as root and as non-root:
> 
> @z3ntu
> what about after [ab2516f](https://github.com/libusb/hidapi/commit/ab2516fcbae280c42911f6704537ff2fe4afd409) ?

I did a quick test removing my udev rules and it didn't work. We have to read directly from the hidraw interface `/dev/hidraw*` so unless udev has given some read access the default looks like this (at least on Arch Linux)

```
crw-------  1 root root 237, 2 Nov 18 13:44 /dev/hidraw2
crw-------  1 root root 237, 3 Nov 18 13:44 /dev/hidraw3
crw-------  1 root root 237, 4 Nov 18 13:44 /dev/hidraw4
```

Youw commented:
in such case I'd rather suggest using a whole different approach:
instead of sending ioctl, read the descriptor from _sysfs_, e.g.: /sys/class/hidraw/_**hidraw0**_/device/**report_descriptor**

Minimum supported kernel 2.6.39 [has the implementation](https://github.com/torvalds/linux/blob/v2.6.39/drivers/hid/hid-core.c#L1182) for this sysfs file, so there will be no backward compatibility issues.

Some thoughts out loud:
In such way, we could also enumerate usage/usage_page for libusb backend on linux.

haata commented:
Yeah, I think this might work well.
Let me put something together for hidraw.

libusb is possible if it's just for Linux (might take some work to map libusb to hidraw; hidraw doesn't make it easy to go in reverse, I've tried)
Something else might have to be done for freebsd (I haven't used BSD that much). But if the hid descriptors are exposed the parsing code would be the same.

haata commented:
A little bit more string manipulation than I would have liked, but I think it works well.

The only part I'm a bit concerned about is during the open where I have to locate the report_descriptor. I'm using the basename of the filepath as the udev sysname to do the sysfs path lookup.
This will break if a symlink is used, so I may also need to resolve any symlinks.
I am open to a better solution though.

z3ntu commented:
hidtest output is now idential as root and non-root with my current devices (and it wasn't on commit ab2516f or before)

Youw commented on linux/hid.c:
```diff
@@ -220,6 +257,141 @@ static int uses_numbered_reports(__u8 *report_descriptor, __u32 size) {
+
+	rpt_handle = open(rpt_path, O_RDONLY);
+	if (rpt_handle == -1) {
+		register_global_error(strerror(errno));
```
let's make it more meaningful with `register_global_error_format`

Youw commented on linux/hid.c:
```diff
@@ -220,6 +257,141 @@ static int uses_numbered_reports(__u8 *report_descriptor, __u32 size) {
+	memset(rpt_desc, 0x0, sizeof(*rpt_desc));
+	res = read(rpt_handle, rpt_desc->value, HID_MAX_DESCRIPTOR_SIZE);
+	if (res == -1) {
+		register_global_error(strerror(errno));
```
ditto

Youw commented on linux/hid.c:
```diff
@@ -582,6 +754,53 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
 			}
+
+			/* Construct <sysfs_path>/device/report_descriptor */
+			char *rpt_path = (char *)calloc(1, strlen(sysfs_path) + 25);
+			sprintf(rpt_path, "%s/device/report_descriptor", sysfs_path);
```
1) Correct me if I'm wrong, but 25 is not enough to hold a NULL-terminator
2) we probably want to use `snprintf` instead

Youw commented on linux/hid.c:
```diff
@@ -582,6 +754,53 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
+					usage_count++;
+
+					/* First usage */
+					if (usage_count == 1)
+					{
+						cur_dev->usage_page = page;
+						cur_dev->usage = usage;
+						continue;
+					}
```
have it only once, outside the loop, and loose `usage_count` variable

haata replied:
Actually I think I have solution.

Youw replied:
so do I, in case if you want to compare :)

Youw commented on linux/hid.c:
```diff
@@ -582,6 +754,53 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
+					cur_dev->path = strdup(dev_path);
+					cur_dev->vendor_id = dev_vid;
+					cur_dev->product_id = dev_pid;
+					cur_dev->serial_number = utf8_to_wchar_t(serial_number_utf8);
```
`wcsdup` from prev dev?

Youw commented on linux/hid.c:
```diff
@@ -672,28 +891,38 @@ hid_device * HID_API_EXPORT hid_open_path(const char *path)
+
 		/* Get the report descriptor */
-		int res, desc_size = 0;
+		int res;
 		struct hidraw_report_descriptor rpt_desc;
 
-		memset(&rpt_desc, 0x0, sizeof(rpt_desc));
-
-		/* Get Report Descriptor Size */
-		res = ioctl(dev->device_handle, HIDIOCGRDESCSIZE, &desc_size);
-		if (res < 0)
-			register_device_error_format(dev, "ioctl (GRDESCSIZE): %s", strerror(errno));
-
-		/* Get Report Descriptor */
-		rpt_desc.size = desc_size;
-		res = ioctl(dev->device_handle, HIDIOCGRDESC, &rpt_desc);
-		if (res < 0) {
-			register_device_error_format(dev, "ioctl (GRDESC): %s", strerror(errno));
-		} else {
+		res = get_hid_report_descriptor(rpt_path, &rpt_desc);
```
Hmm, I agree that a bit more string manipulations than than one would wan't to see.
And yes, based on my experience, using `realpath` is a good call.

Since at this point we already have a device handle, I suggest to keep the `ioctl` approach for this scenario: it looks a bit cleaner, and less error prone (as in case with `25`, in my other comment).

haata replied:
Agreed.

Youw reviewed (commented):
I didn't realize we had a `sysfs_path` already available from udev, during enumeration.
This makes the implementation even simpler. I do like how it looks.
But still, have a few comments below.

Youw commented on linux/hid.c:
```diff
@@ -582,6 +754,53 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
+					cur_dev->path = strdup(dev_path);
+					cur_dev->vendor_id = dev_vid;
+					cur_dev->product_id = dev_pid;
+					cur_dev->serial_number = wcsdup(prev_dev->serial_number);
```
```suggestion
					cur_dev->serial_number = prev_dev->serial_number? wcsdup(prev_dev->serial_number): NULL;
```

Youw commented on linux/hid.c:
```diff
@@ -582,6 +754,53 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
 			}
+
+			/* Construct <sysfs_path>/device/report_descriptor */
+			int rpt_path_len = strlen(sysfs_path) + 25 + 1;
```
```suggestion
			size_t rpt_path_len = strlen(sysfs_path) + 25 + 1;
```

Youw reviewed (approved):
Nicely done!

I'd like to check it myself with some of my devices (over the weekend).

haata commented:
> I'd like to check it myself with some of my devices (over the weekend).

I'll do some additional checking myself too. Bluetooth devices tend to have more corner cases as they tend to favour multiple usage/usage page pairs instead of multiple devices.

Should I squash all of the commits together?

Youw commented:
Not necessary, I'd do a squash-merge anyway.

vincele commented on linux/hid.c:
```diff
@@ -220,6 +256,157 @@ static int uses_numbered_reports(__u8 *report_descriptor, __u32 size) {
+			break;
+		}
+
+		/* Skip over this key and it's associated data */
```
it's "its"

Youw replied:
https://github.com/libusb/hidapi/pull/368
